File: backend-go/internal/data/embed_hash.go
package data

import (
	"crypto/sha256"
	"fmt"
	"math"
	"strings"
)

const defaultEmbedDim = 384

// HashEmbedJSON returns normalized hash embedding JSON for sqlite-vec (migrate parity).
func HashEmbedJSON(text string, dim int) string {
	if dim <= 0 {
		dim = defaultEmbedDim
	}
	vec := hashEmbed(text, dim)
	vals := make([]string, dim)
	for i, v := range vec {
		vals[i] = fmt.Sprintf("%g", v)
	}
	return "[" + strings.Join(vals, ",") + "]"
}

// HashEmbed returns normalized hash embedding vector (migrate parity).
func HashEmbed(text string, dim int) []float32 {
	return hashEmbed(text, dim)
}

func hashEmbed(text string, dim int) []float32 {
	out := make([]float32, dim)
	for i := 0; i < dim; i++ {
		h := sha256.Sum256([]byte(fmt.Sprintf("%s:%d", text, i)))
		out[i] = (float32(int(h[0])<<8|int(h[1]))/65535)*2 - 1
	}
	var norm float64
	for _, v := range out {
		norm += float64(v) * float64(v)
	}
	norm = math.Sqrt(norm)
	if norm > 0 {
		for i := range out {
			out[i] = float32(float64(out[i]) / norm)
		}
	}
	return out
}

// CosineSimilarity between two equal-length vectors.
func CosineSimilarity(a, b []float32) float64 {
	if len(a) == 0 || len(a) != len(b) {
		return 0
	}
	var dot, na, nb float64
	for i := range a {
		dot += float64(a[i]) * float64(b[i])
		na += float64(a[i]) * float64(a[i])
		nb += float64(b[i]) * float64(b[i])
	}
	if na == 0 || nb == 0 {
		return 0
	}
	return dot / (math.Sqrt(na) * math.Sqrt(nb))
}

File: backend-rust/src/data/embed_hash.rs
//! Normalized hash embeddings — mirrors `backend-go/internal/data/embed_hash.go`.

use sha2::{Digest, Sha256};

pub const DEFAULT_EMBED_DIM: usize = 384;

pub fn hash_embed(text: &str, dim: usize) -> Vec<f32> {
    let dim = if dim == 0 { DEFAULT_EMBED_DIM } else { dim };
    let mut out = vec![0.0f32; dim];
    for i in 0..dim {
        let digest = Sha256::digest(format!("{text}:{i}"));
        let v = (digest[0] as u32) << 8 | digest[1] as u32;
        out[i] = (v as f32 / 65535.0) * 2.0 - 1.0;
    }
    let mut norm = 0.0f64;
    for v in &out {
        norm += f64::from(*v) * f64::from(*v);
    }
    let norm = norm.sqrt();
    if norm > 0.0 {
        for v in &mut out {
            *v = (*v as f64 / norm) as f32;
        }
    }
    out
}

pub fn cosine_similarity(a: &[f32], b: &[f32]) -> f64 {
    if a.is_empty() || a.len() != b.len() {
        return 0.0;
    }
    let mut dot = 0.0f64;
    let mut na = 0.0f64;
    let mut nb = 0.0f64;
    for i in 0..a.len() {
        dot += f64::from(a[i]) * f64::from(b[i]);
        na += f64::from(a[i]) * f64::from(a[i]);
        nb += f64::from(b[i]) * f64::from(b[i]);
    }
    if na == 0.0 || nb == 0.0 {
        return 0.0;
    }
    dot / (na.sqrt() * nb.sqrt())
}
